topostats/statistics.py

```diff
@@ -0,0 +1,81 @@
+"""Function for calculating statistics about a whole image, for example number of grains or surface roughness."""
+
+import numpy as np
+import pandas as pd
+
+from topostats.logs.logs import setup_logger, LOGGER_NAME
+from topostats.roughness import roughness_rms
+
+LOGGER = setup_logger(LOGGER_NAME)
+
+
+def image_statistics(
+    image: np.ndarray,
+    filename: str,
+    pixel_to_nm_scaling: float,
+    results_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """Calculate statistics pertaining to the whole image, for example the size of the image in pixels and
+    metres, the root-mean-squared roughness and the grains per metre squared.
+
+    Parameters
+    ----------
+    image: np.ndarray
+        Numpy 2D image array of the image to calculate stats for.
+    filename: str
+        The name of the file being processed.
+    pixel_to_nm_scaling: float
+        Float of the scaling factor between pixels and nanometres.
+    results_df: pd.DataFrame
+        Pandas DataFrame of statistics pertaining to individual grains including from grainstats and
+        dna tracing.
+
+    Returns
+    -------
+    dict
+        Dictionary of image statistics.
+    """
+
+    image_stats = {
+        "Image": filename,
```

Comment:
I don't think you need to add the `Image` to the data frame here because you've changed the loop in `run_topostats.py` to save the results into a dictionary which when you `pd.concat` uses the dictionary index as the index for the dataframe.

Might be worth checking but I think this is what happens with the grain/tracing statistics.

Reply:
I think it does need it. I tried removing it and...

![image](https://github.com/AFM-SPM/TopoStats/assets/86117496/8a758576-a00f-419c-9b8a-d00dfd7a91f6)

`grainstats.py` also includes the image name:
```
        else:
            grainstats_df = create_empty_dataframe()
        grainstats_df.index.name = "molecule_number"
        grainstats_df["image"] = self.image_name

        return grainstats_df, grains_plot_data
```

Reply:
You are of course right, the `image_name` is used in the dictionary which collects results for the parallel processing. We only use the `.values()` of that dictionary when passing to `pd.concat()` though since it only works with objects and doesn't care for the `.keys()`.